### PR TITLE
Pick most specific matching virtual service in mesh for TCP/TLS

### PR DESCRIPTION
**Pull request by ijsnellf**
Right now the code picks the first matching virtual service. Since virtual services are sorted by creation time, it uses the oldest matching virtual service. This PR fixes the logic to use the virtual service that matches most specifically.

### CODE REVIEW

File: pilot/pkg/networking/core/v1alpha3/tls_test.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+import (
+	"testing"
+
+	"istio.io/api/networking/v1alpha3"
+	"istio.io/istio/pilot/pkg/model"
+)
+
+func TestVirtualServiceForHost(t *testing.T) {
+	wildcard := &v1alpha3.VirtualService{Hosts: []string{"*"}}
```
Comment (ijsnellf) [inline pilot/pkg/networking/core/v1alpha3/tls_test.go]: * is not valid in mesh. removing...

File: pilot/pkg/networking/core/v1alpha3/tls.go [hunk 2]
```diff
@@ -247,10 +235,11 @@ func buildSidecarOutboundTCPFilterChainOpts(node *model.Proxy, env *model.Enviro
 	return out
 }
 
-func buildSidecarOutboundTCPTLSFilterChainOpts(node *model.Proxy, env *model.Environment, configs []model.Config, destinationIPAddress string,
-	service *model.Service, listenPort *model.Port, proxyLabels model.LabelsCollection, gateways map[string]bool) []*filterChainOpts {
+func buildSidecarOutboundTCPTLSFilterChainOpts(node *model.Proxy, env *model.Environment, destinationIPAddress string,
+	service *model.Service, listenPort *model.Port, proxyLabels model.LabelsCollection, gateways map[string]bool,
+	byHost virtualServiceForHost) []*filterChainOpts {
```
Comment (rshriram) [inline pilot/pkg/networking/core/v1alpha3/tls.go]: time to make this a context variable where we stuff all args?

Comment (ijsnellf) [reply pilot/pkg/networking/core/v1alpha3/tls.go]: ...maybe.

File: pilot/pkg/networking/core/v1alpha3/tls.go [hunk 3]
```diff
@@ -247,10 +235,11 @@ func buildSidecarOutboundTCPFilterChainOpts(node *model.Proxy, env *model.Enviro
...
 
-	virtualService := getVirtualServiceForHost(service.Hostname, configs)
+	virtualService := byHost(service.Hostname)
```
Comment (rshriram) [inline pilot/pkg/networking/core/v1alpha3/tls.go]: Can we keep this in the simple function call way and not pass functions as parameters and have implicit variables in the functions? Its such a pain to follow the code.

File: pilot/pkg/networking/core/v1alpha3/tls.go [hunk 4]
```diff
@@ -260,3 +249,31 @@ func buildSidecarOutboundTCPTLSFilterChainOpts(node *model.Proxy, env *model.Env
...
+		vs := config.Spec.(*v1alpha3.VirtualService)
+		if vs != nil {
+			for _, vsHost := range vs.Hosts {
+				hostname := model.Hostname(vsHost)
+				hostnames = append(hostnames, hostname)
+				byHostname[hostname] = vs
+			}
+		}
+	}
+	sort.Sort(hostnames) // order from most to least specific
```
Comment (rshriram) [inline pilot/pkg/networking/core/v1alpha3/tls.go]: where is the sort function?

File: pilot/pkg/networking/core/v1alpha3/tls.go [hunk 5]
```diff
@@ -260,3 +249,31 @@ func buildSidecarOutboundTCPTLSFilterChainOpts(node *model.Proxy, env *model.Env
...
+	return func(hostname model.Hostname) *v1alpha3.VirtualService {
+		for _, vsHost := range hostnames {
+			if vsHost.Matches(hostname) {
+				return byHostname[vsHost] // first match is the most specific
+			}
+		}
+		return nil
+	}
+}
```
Comment (rshriram) [inline pilot/pkg/networking/core/v1alpha3/tls.go]: this code can move into the model.. Like model.VirtualServicesForHost() or something.
And while this code is slightly optimized, moving it to model will allow us to optimize this more, by creating that map once in the beginning and using it for all sidecars. See costin's PR for example.

File: pilot/pkg/networking/core/v1alpha3/tls.go (hunk 4 above)
Comment (ijsnellf) [reply pilot/pkg/networking/core/v1alpha3/tls.go]: `model.Hostnames` implements the sorting interface and sorts with most to least specific ordering.

Comment (ijsnellf) [reply pilot/pkg/networking/core/v1alpha3/tls.go]: https://github.com/istio/istio/blob/master/pilot/pkg/model/service.go#L524

**Comment by ijsnellf** [comment]
Closing this in favor of adding a `VirtualServicesForHost` to the push context being built here: https://github.com/istio/istio/pull/7635

**Pull request closed without merging**